lib/welcome/pages/login_page.dart

```diff
@@ -0,0 +1,166 @@
+import 'dart:core';
+
+import 'package:flutter/material.dart';
+import 'package:flutter_app/util/email_field.dart';
+import 'package:flutter_app/util/loading_button.dart';
+import 'package:flutter_app/util/password_field.dart';
+import 'package:flutter_app/welcome/pages/forgot_password_page.dart';
+import 'package:flutter_app/welcome/pages/register_page.dart';
+import 'package:flutter_app/util/supabase.dart';
+import 'package:progress_state_button/progress_button.dart';
+import 'package:supabase_flutter/supabase_flutter.dart';
+
+class LoginPage extends StatefulWidget {
+  const LoginPage({super.key});
+
+  @override
+  State<LoginPage> createState() => _LoginPageState();
+}
+
+class _LoginPageState extends State<LoginPage> {
+  @override
+  Widget build(BuildContext context) {
+    return Scaffold(
+      appBar: AppBar(
+        title: const Text("Login"),
+      ),
+      body: const Center(
+        child: CustomScrollView(
+          physics: ClampingScrollPhysics(),
+          slivers: [
+            SliverFillRemaining(
+              hasScrollBody: false,
+              child: Padding(
+                padding: EdgeInsets.symmetric(horizontal: 40, vertical: 20),
+                child: LoginForm(),
+              ),
+            )
+          ],
+        ),
+      ),
+    );
+  }
+}
+
+class LoginForm extends StatefulWidget {
+  const LoginForm({super.key});
+
+  @override
+  State<LoginForm> createState() => _LoginFormState();
+}
+
+class _LoginFormState extends State<LoginForm> {
+  final GlobalKey<FormState> _formKey = GlobalKey<FormState>();
+  final emailController = TextEditingController();
+  final passwordController = TextEditingController();
+  ButtonState _state = ButtonState.idle;
+
+  void onSubmit() async {
+    if (_formKey.currentState!.validate()) {
+      setState(() {
+        _state = ButtonState.loading;
+      });
+      try {
+        await supabaseClient.auth.signInWithPassword(
+          email: emailController.text,
+          password: passwordController.text,
+        );
+        setState(() {
+          _state = ButtonState.success;
+        });
+      } on AuthException catch (e) {
+        fail();
+        // looks weird but needed later for i18n
+        String text = e.statusCode == '400'
+            ? (e.message.contains("credentials")
+                ? "Invalid credentials"
+                : "Please confirm your email address")
+            : "Something went wrong";
+
+        ScaffoldMessenger.of(context).showSnackBar(SnackBar(
+          content: Text(text),
+        ));
+      }
+    } else {
+      fail();
+    }
+  }
+
+  void fail() async {
```

Comment:
can we refactor this method? because it's the same everywhere it's used.

Reply:
sadly, we don't really think there's a good way to do that :( Guess we'll have to accept the duplication.